### ISSUE TITLE
testing fizz buzz

### ISSUE THREAD

**Issue by ivoysey** [jasonyih/TestRepo#1]
Right now, it's not immediately clear to me how I would test the code you wrote to see if it's correct. It prints things to standard out, which is great, but that makes it hard to run test cases: you'd have to capture the strings it produces and then compare them to some other strings, and it'd be really messy.

Can you think of a way to restructure this code so that it's testable? If I had another implementation of `fizzbuzz`, for example, and wanted to say "yes they do the same thing for these inputs", how could we do that?

**Comment by jasonyih** [comment]
Maybe to make it testable, I could create a whole different fizzBuzz method
and add a Scanner to accept input from the user. Then I could change it so
that the method returns a String instead of just printing the numbers. I
think it would make it easier to compare my implementation of the method to
another one since we could compare the Strings returned by the methods.

**Comment by ivoysey** [comment]
OK! That's one way to do it, for sure. Another thing you could do is return an array of strings from any implementation of fizzbuzz and then have another method that just does the printing but none of the math.

Maybe try it both ways and see which one feels better?

**Comment by jasonyih** [comment]
I tried both ways, and returning an array definitely feels like the better
way. It makes everything more organized, and it also makes it much easier
to test because there isn't any formatting in the array. Returning a String
with all of the formatting would make it very hard to test because
different implementations of the method could have different formatting, so
returning the array of Strings is better because a correct implementation
of the method will return the same array.

**Comment by ivoysey** [comment]
I thought it might end up feeling that way. Commit your attempts, though! "Commit early and often" is the right way to think about this. In my mind it's always better to commit code that might not be perfect but has made a change than to just wait and wait and wait.

**Comment by jasonyih** [comment]
Oh I thought I committed my new file, or does adding "files via upload" not
count as a commit? I guess I'm still not completely sure about when to
commit my changes.

**Comment by ivoysey** [comment]
What do you mean by "adding files by upload"? Is that a github feature?

Commit is a verb for a git action that says "mark down all of the changes
i've made so far and call that the new current state of the repo". You may
not have pushed your changes to the github repo and only made them locally
-- push is another, separate, git action so you can make a bunch of small
changes and keep track of them but not share them with everyone else until
you're ready.

**Comment by jasonyih** [comment]
Yeah, on Github I uploaded the new version of my Java program, and the
description was "Add files via upload." I think the problem might be that I
didn't make a Git repository for the Java project, and I was doing things
solely on Github. I think I should probably create a local repository and
connect it to Github?

**Comment by ivoysey** [comment]
Ah, yes. You definintely want a local repo that has github set as upstream master. My flow for this is usually that I make a repo on github first with just a `README.md` (or a license or whatever, it just can't be totally empty), and then clone that to my machine. You can of course do `git init` and then `git --set-`...something I can never remember. So I usually don't force myself to google that by needing to remember it. :)

**Comment by jasonyih** [comment]
I see now, that makes a lot more sense! I cloned the TestRepo to my
computer and then it was automatically connected to github. Now I can just
make changes locally then commit them locally as well, without interfering
with the Github repo. I also tested pushing the changes onto the online
repo, and it worked instantly. Are there any other things I should test out
with connecting Git and Github repos?

**Comment by ivoysey** [comment]
Probably, but we can figure them out as we discover them. 😄

**Comment by jasonyih** [comment]
Sounds good, thanks!
